### PR TITLE
PerformanceCounterLayoutRenderer - Added Layout support for Instance

### LINKED ISSUE #3569: Performance counter layout renderer parse bug
**Nlog version**: 4.6.7

**Platform**: .Net 4.7

**Current NLog config** (xml or C#, if relevant)

```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" internalLogFile="c:/logs/nlog.log" internalLogLevel="Trace">
    <variable name="facilityName" value="BppprCalc" />
    <targets async="true">
        <target name="graylogCounters" xsi:type="GraylogHttp" facility="${facilityName}" graylogServer="http://graylog" graylogPort="12201">
        <parameter name="processor_total" layout="${performancecounter:Category=Processor:Counter=% Processor Time:Instance=_Total}" />
        <parameter name="requests_per_second" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Requests/Sec:Instance=${gdc:InstanceName}}" />
        <parameter name="requests_in_queue" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Requests in Application Queue:Instance=${gdc:InstanceName}}" />
        <parameter name="errors_total_per_second" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Errors Total/Sec:Instance=${gdc:InstanceName}}" />
      </target>
    </targets>
    <rules>
      <logger name="counters" level="debug" writeTo="graylogCounters" final="true"/>
    </rules>
  </nlog>
```

- What is the current result?
Layout performance counter gives an error while parsing(I think) **gdc:**. I have putted variables in gdc and value is all right.
This is internal logs of setup.
```
2019-08-30 13:19:16.5796 Warn Exception in layout renderer. Exception: System.InvalidOperationException: Экземпляр '${gdc' не существует в указанной категории.
  в System.Diagnostics.CounterDefinitionSample.GetInstanceValue(String instanceName)
  в System.Diagnostics.PerformanceCounter.NextSample()
  в NLog.LayoutRenderers.PerformanceCounterLayoutRenderer.GetValue()
  в NLog.LayoutRenderers.PerformanceCounterLayoutRenderer.Append(StringBuilder builder, LogEventInfo logEvent)
  в NLog.LayoutRenderers.LayoutRenderer.RenderAppendBuilder(LogEventInfo logEvent, StringBuilder builder)
``` 

In english the result is the gdc do not exist in selected category.
- What is the expected result?
performance counters layout renderer works well and can get data from gdc.
- I also tried to add a variable into target and reconfigure my config at application startup(global.asax).

```xml
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" internalLogFile="c:/logs/nlog.log" internalLogLevel="Trace">
    <variable name="facilityName" value="BppprCalc" />
    <variable name="instanceName" value="DefaultInstance" />
    <targets async="true">
      <target name="graylogCounters" xsi:type="GraylogHttp" facility="${facilityName}" graylogServer="http://graylog" graylogPort="12201">
        <parameter name="processor_total" layout="${performancecounter:Category=Processor:Counter=% Processor Time:Instance=_Total}" />
        <parameter name="requests_per_second" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Requests/Sec:Instance=${instanceName}}" />
        <parameter name="requests_in_queue" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Requests in Application Queue:Instance=${instanceName}}" />
        <parameter name="errors_total_per_second" layout="${performancecounter:Category=ASP.NET Apps v4.0.30319:Counter=Errors Total/Sec:Instance=${instanceName}}" />
      </target>
    </targets>
    <rules>
      <logger name="counters" level="debug" writeTo="graylogCounters" final="true"/>
    </rules>
  </nlog>
``` 
code at startup
```c#
NLog.LogManager.Configuration.Variables["instanceName"] = "_LM_W3SVC_1_ROOT_BppprCalc";
NLog.LogManager.ReconfigExistingLoggers();
```
and there is another mistake

```
 2019-08-30 14:04:12.7192 Warn Exception in layout renderer. Exception: System.InvalidOperationException: Экземпляр 'DefaultInstance' не существует в указанной категории.
    в System.Diagnostics.CounterDefinitionSample.GetInstanceValue(String instanceName)
    в System.Diagnostics.PerformanceCounter.NextSample()
    в NLog.LayoutRenderers.PerformanceCounterLayoutRenderer.GetValue()
    в NLog.LayoutRenderers.PerformanceCounterLayoutRenderer.Append(StringBuilder builder, LogEventInfo logEvent)
    в NLog.LayoutRenderers.LayoutRenderer.RenderAppendBuilder(LogEventInfo logEvent, StringBuilder builder)
```

### PR DESCRIPTION
**Pull request by snakefoot**
Resolves #3569

Note merge to master

### CODE REVIEW

**Comment by 304NotModified** [comment]
@snakefoot please do a force push, that should fix Azure pipelines

**Comment by 304NotModified** [comment]
added merge, also for triggering azure pipelines

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs [hunk 1]
```diff
@@ -71,13 +71,33 @@ public class PerformanceCounterLayoutRenderer : LayoutRenderer
...
         /// <docgen category='Performance Counter Options' order='10' />
-        public string MachineName { get; set; }
+        public string MachineName
+        {
+            get { return _machineName?.Text; }
+            set
+            {
+                _machineName = new SimpleLayout(value);
+                _fixedPerformanceCounter = null;
+                _performanceCounter = null;
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: Please create a helper `ClearFields` (3 times needed), to set them all to `null`

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs [hunk 2]
```diff
@@ -96,24 +116,67 @@ protected override void InitializeLayoutRenderer()
...
+            _performanceCounter?.Close();
+            _performanceCounter = null;
+        }
+
+        /// <inheritdoc />
+        protected override void Append(StringBuilder builder, LogEventInfo logEvent)
+        {
+            var performanceCounter = LookupPerformanceCounter(logEvent);
+            var formatProvider = GetFormatProvider(logEvent, Culture);
+            builder.Append(performanceCounter.GetValue().ToString(Format, formatProvider));
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: Couldn't .GetValue return `null`?

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs [hunk 3]
```diff
@@ -152,48 +215,51 @@ private static string GetCurrentProcessInstanceName(string category)
             return string.Empty;
         }
 
-        /// <inheritdoc />
-        protected override void CloseLayoutRenderer()
+        class PerformanceCounterCached
         {
-            base.CloseLayoutRenderer();
-            if (_perfCounter != null)
+            public PerformanceCounterCached(string machineName, string instanceName, PerformanceCounter performanceCounter)
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: I prefer a consistent ordering, 

See https://stackoverflow.com/a/310967/201303

Thanks!

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: `GetValue()` returns float, so it cannot be null.

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: Fixed

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: Fixed

File: src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/LayoutRenderers/PerformanceCounterLayoutRenderer.cs]: :+1:

**Comment by 304NotModified** [comment]
Thanks!

**Comment by 304NotModified** [comment]
@snakefoot Do you think we should document this somewhere?

**Comment by snakefoot** [comment]
> @304NotModified Do you think we should document this somewhere?

Well everyone think that all properties in NLog supports layoutrenderers, then I guess it is okay giving the users that illusion. Instead of writing everywhere that it doesn't work.

**Comment by 304NotModified** [comment]
OK!

**Pull request merged**